File: client/src/services/patientService.ts
import api from './api';
import type { ApiResponse } from '../types/api';

export interface Patient {
  id: string;
  clinicId: string;
  patientCode: string;
  fullName: string;
  phoneNumber: string;
  email?: string;
  dateOfBirth?: string;
  address?: string;
  gender?: string;
  allergies?: string;
  chronicConditions?: string;
  emergencyContactName?: string;
  emergencyContactPhone?: string;
  bloodType?: string;
  idNumber?: string;
  insuranceNumber?: string;
  insuranceProvider?: string;
  occupation?: string;
  referralSource?: string;
  firstVisitDate?: string;
  receivePromotions: boolean;
  notes?: string;
  createdAt: string;
}

export interface CreatePatientRequest {
  fullName: string;
  phoneNumber: string;
  email?: string;
  dateOfBirth?: string;
  address?: string;
  gender?: string;
  allergies?: string;
  chronicConditions?: string;
  emergencyContactName?: string;
  emergencyContactPhone?: string;
  bloodType?: string;
  idNumber?: string;
  insuranceNumber?: string;
  insuranceProvider?: string;
  occupation?: string;
  referralSource?: string;
  receivePromotions?: boolean;
  notes?: string;
}

export interface PatientsListResponse {
  items: Patient[];
  total: number;
  page: number;
  pageSize: number;
}

export const patientService = {
  getPatients: async (search?: string, page = 1, pageSize = 10): Promise<ApiResponse<PatientsListResponse>> => {
    const params = new URLSearchParams();
    if (search) params.append('search', search);
    params.append('page', page.toString());
    params.append('pageSize', pageSize.toString());
    
    const response = await api.get<ApiResponse<PatientsListResponse>>(`/patients?${params}`);
    return response.data;
  },

  getPatient: async (id: string): Promise<ApiResponse<Patient>> => {
    const response = await api.get<ApiResponse<Patient>>(`/patients/${id}`);
    return response.data;
  },

  createPatient: async (data: CreatePatientRequest): Promise<ApiResponse<Patient>> => {
    const response = await api.post<ApiResponse<Patient>>('/patients', data);
    return response.data;
  },

  updatePatient: async (id: string, data: CreatePatientRequest): Promise<ApiResponse<Patient>> => {
    const response = await api.put<ApiResponse<Patient>>(`/patients/${id}`, data);
    return response.data;
  },

  deletePatient: async (id: string): Promise<ApiResponse<void>> => {
    const response = await api.delete<ApiResponse<void>>(`/patients/${id}`);
    return response.data;
  },
};

File: src/ClinicManagement.Infrastructure/Services/PatientService.cs
using ClinicManagement.Core.Entities;
using ClinicManagement.Core.Interfaces;

namespace ClinicManagement.Infrastructure.Services;

public class PatientService : IPatientService
{
    private readonly IPatientRepository _patientRepository;

    public PatientService(IPatientRepository patientRepository)
    {
        _patientRepository = patientRepository;
    }

    public async Task<(bool Success, string? ErrorCode, Patient? Patient)> CreatePatientAsync(
        Guid clinicId,
        string fullName,
        string phoneNumber,
        string? email,
        DateTime? dateOfBirth,
        string? address,
        string? gender,
        string? allergies,
        string? chronicConditions,
        string? emergencyContactName,
        string? emergencyContactPhone,
        string? bloodType,
        string? idNumber,
        string? insuranceNumber,
        string? insuranceProvider,
        string? occupation,
        string? referralSource,
        bool receivePromotions,
        string? notes)
    {
        var clinicPatientCount = await _patientRepository.GetClinicPatientCountAsync(clinicId);
        var patientCount = clinicPatientCount + 1;

        var patient = new Patient
        {
            ClinicId = clinicId,
            PatientCode = $"PT{patientCount:D5}",
            FullName = fullName,
            PhoneNumber = phoneNumber,
            Email = email,
            DateOfBirth = dateOfBirth,
            Address = address,
            Gender = gender,
            Allergies = allergies,
            ChronicConditions = chronicConditions,
            EmergencyContactName = emergencyContactName,
            EmergencyContactPhone = emergencyContactPhone,
            BloodType = bloodType,
            IdNumber = idNumber,
            InsuranceNumber = insuranceNumber,
            InsuranceProvider = insuranceProvider,
            Occupation = occupation,
            ReferralSource = referralSource,
            FirstVisitDate = DateTime.UtcNow,
            ReceivePromotions = receivePromotions,
            Notes = notes
        };

        await _patientRepository.AddAsync(patient);
        await _patientRepository.SaveChangesAsync();

        return (true, null, patient);
    }

    public async Task<(bool Success, string? ErrorCode, Patient? Patient)> UpdatePatientAsync(
        Guid id,
        string fullName,
        string phoneNumber,
        string? email,
        DateTime? dateOfBirth,
        string? address,
        string? gender,
        string? allergies,
        string? chronicConditions,
        string? emergencyContactName,
        string? emergencyContactPhone,
        string? bloodType,
        string? idNumber,
        string? insuranceNumber,
        string? insuranceProvider,
        string? occupation,
        string? referralSource,
        bool receivePromotions,
        string? notes)
    {
        var patient = await _patientRepository.GetByIdAsync(id);
        if (patient == null)
        {
            return (false, "NOT_FOUND", null);
        }

        patient.FullName = fullName;
        patient.PhoneNumber = phoneNumber;
        patient.Email = email;
        patient.DateOfBirth = dateOfBirth;
        patient.Address = address;
        patient.Gender = gender;
        patient.Allergies = allergies;
        patient.ChronicConditions = chronicConditions;
        patient.EmergencyContactName = emergencyContactName;
        patient.EmergencyContactPhone = emergencyContactPhone;
        patient.BloodType = bloodType;
        patient.IdNumber = idNumber;
        patient.InsuranceNumber = insuranceNumber;
        patient.InsuranceProvider = insuranceProvider;
        patient.Occupation = occupation;
        patient.ReferralSource = referralSource;
        patient.ReceivePromotions = receivePromotions;
        patient.Notes = notes;

        await _patientRepository.UpdateAsync(patient);
        await _patientRepository.SaveChangesAsync();

        return (true, null, patient);
    }

    public async Task<(bool Success, string? ErrorCode)> DeletePatientAsync(Guid id)
    {
        var patient = await _patientRepository.GetByIdAsync(id);
        if (patient == null)
        {
            return (false, "NOT_FOUND");
        }

        await _patientRepository.SoftDeleteAsync(patient);
        await _patientRepository.SaveChangesAsync();

        return (true, null);
    }

    public async Task<Patient?> GetPatientByIdAsync(Guid id)
    {
        return await _patientRepository.GetByIdAsync(id);
    }

    public async Task<(IEnumerable<Patient> Items, int Total)> SearchPatientsAsync(string? search, int page, int pageSize)
    {
        var items = await _patientRepository.SearchPatientsAsync(search, page, pageSize);
        var total = await _patientRepository.GetTotalCountAsync(search);
        return (items, total);
    }
}